### ISSUE TITLE
API Design Quality Report: Consistency and Abstraction Analysis

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tensor4all-rs#408]
# API Design Quality Report

## Executive Summary

The tensor4all-rs codebase demonstrates **strong Rust idioms** with thoughtful abstraction, but exhibits several **consistency issues** that could be improved. The design shows clear influence from ITensors.jl while adapting to Rust's type system.

---

### 1.1 Well-Structured Trait Hierarchy

The `IndexLike` trait provides a clean abstraction for index types:

```rust
pub trait IndexLike {
    fn id(&self) -> &Self::Id;
    fn dim(&self) -> usize;
    fn plev(&self) -> i64;
    fn conj_state(&self) -> ConjState;
    fn conj(&self) -> Self;
    fn sim(&self) -> Self;  // "similar" - creates equivalent with new ID
}
```

**Why this works:** Default implementations for `is_contractable()`, `same_id()`, and `has_id()` reduce boilerplate while allowing customization.

### 1.2 Consistent Constructor Patterns

The codebase follows Rust conventions:

| Pattern | Example | Purpose |
|---------|---------|----------|
| `new()` | `TensorTrain::new(tensors)` | Primary constructor with validation |
| `from_*()` | `SiteTensorTrain::from_tensor_train(tt, center)` | Conversion from related types |
| Domain-specific | `TensorTrain::zeros(site_dims)`, `constant(site_dims, value)` | Common special cases |

### 1.3 Thoughtful Method Naming for Tensor Operations

```rust
// Clear semantic distinction
tt.contract(other)        // Auto-detect common indices
tt.tensordot(other, pairs) // Explicit index pairing (NumPy-style)
tt.outer_product(other)   // No contraction
```

---

### 2.1 Inconsistent Return Types for Similar Operations

**Issue:** Some methods return `Result<T>`, others panic or return `Option<T>`:

```rust
// Returns Result - good
pub fn evaluate(&self, indices: &[LocalIndex]) -> Result<T>

// Panics on error - inconsistent  
pub fn site_dim(&self, i: usize) -> usize  // No bounds check visible

// Returns Option - different style again
pub fn get(&self, index: &[usize]) -> Option<&T>  // ColMajorArray
```

**Recommendation:** Standardize on `Result<T>` for fallible operations with descriptive error types.

### 2.2 Mixed Naming Conventions for Accessors

| Type | Pattern | Example |
|------|---------|----------|
| TensorTrain | Singular | `site_tensor(i)`, `link_dim(i)` |
| MPO | Plural + singular | `site_tensors()`, `site_tensor(i)` |
| Some types | With "get" prefix | `get(&self, row, col)` in BlockTensor |

**Recommendation:** Adopt consistent pattern: `foo()` for collection, `foo(index)` for element access.

### 2.3 Inconsistent Use of `&mut self` vs Returning New Values

```rust
// Mutable in-place
pub fn compress(&mut self, options: &CompressionOptions) -> Result<()>
pub fn scale(&mut self, factor: T)

// Returns new value  
pub fn compressed(&self, options: &CompressionOptions) -> Result<Self>
pub fn scaled(&self, factor: T) -> Self
```

**Assessment:** This is actually **good design** - follows Rust convention of offering both patterns. However, documentation could clarify when to use each.

---

### 3.1 Trait Bound Proliferation

The scalar trait system has become complex:

```rust
// Multiple overlapping traits for scalars
trait TTScalar
trait Scalar  
trait LinalgScalar
trait SVDCompressScalar
trait EinsumScalar
```

**Problem:** Users must understand which trait to bound on. The relationships aren't clearly documented.

**Recommendation:** Consolidate into a hierarchy or provide a comprehensive "scalar traits" module documentation.

### 3.2 Incomplete Abstraction in `AbstractTensorTrain`

The trait has default implementations that may hide important differences:

```rust
pub trait AbstractTensorTrain {
    fn link_dims(&self) -> Vec<usize>;  // Default impl - may be inefficient
    fn evaluate(&self, indices: &[LocalIndex]) -> Result<T>;  // Default impl
}
```

**Issue:** Default implementations can mask performance differences between `TensorTrain`, `VidalTensorTrain`, and `SiteTensorTrain`.

### 3.3 Type-Specific vs Generic API Inconsistency

```rust
// Generic version (preferred)
pub fn to_vec(&self) -> Result<Vec<T>>

// Type-specific wrappers (legacy?)
pub fn as_slice_f64(&self) -> Result<Vec<f64>>
pub fn as_slice_c64(&self) -> Result<Vec<Complex64>>
```

**Assessment:** The doc comment says "Prefer the generic... This wrapper is kept for C API compatibility." This is acceptable but should be clearly marked as `#[deprecated]` or internal.

---

### 4.1 Feature Creep in Core Types

`TensorDynLen` has accumulated many responsibilities:
- Tensor storage and indexing
- Automatic differentiation (`requires_grad`, `backward`, `grad`)
- Multiple construction paths (`from_dense`, `from_diag`, `copy_tensor`, etc.)

**Recommendation:** Consider splitting AD functionality into a separate crate or module.

### 4.2 Inconsistent Error Type Exposure

```rust
// Some errors are well-typed
pub fn qr(...) -> Result<..., QrError>
pub fn svd(...) -> Result<..., SvdError>

// Others use generic anyhow::Result  
pub fn contract_fit(...) -> Result<MPO<T>>  // What's the error type?
```

**Recommendation:** Use `thiserror` consistently for public APIs.

### 4.3 "Options" Struct Proliferation Without Consistency

| Options Type | Fields | Construction |
|--------------|--------|--------------|
| `CompressionOptions` | tolerance, max_bond_dim, method | `Default` |
| `SvdOptions` | rtol, max_rank | `with_rtol()`, `with_max_rank()` builders |
| `QrOptions` | rtol | `with_rtol()` builder |

**Issue:** No consistent pattern for options construction.

---

### 5.1 The `prime()`/`noprime()` Pattern

```rust
pub fn prime(&self) -> Self      // Increment prime level
pub fn noprime(&self) -> Self    // Reset to zero  
pub fn set_plev(&self, plev: i64) -> Self  // Set explicitly
```

Clean, discoverable API for index manipulation.

### 5.2 Two-Phase Construction for Safety

```rust
// Validation on construction
pub fn new(blocks: Vec<T>, shape: (usize, usize)) -> Result<Self>

// Unsafe variant clearly marked internal  
pub(crate) fn from_tensors_unchecked(tensors: Vec<Tensor3<T>>) -> Self
```

---

## 6. Recommendations Summary

| Priority | Issue | Action |
|----------|-------|--------|
| **High** | Error type consistency | Audit all `Result` returns, use `thiserror` uniformly |
| **High** | Trait bound documentation | Create "Choosing Scalar Traits" guide |
| **Medium** | Accessor naming convention | Standardize on `foo()`/`foo(index)` pattern |
| **Medium** | Options API consistency | Adopt builder pattern or `Default` + overrides uniformly |
| **Low** | Type-specific method cleanup | Mark legacy f64/c64 methods as internal/deprecated |

---

## 7. Overall Assessment

| Aspect | Rating | Notes |
|--------|--------|-------|
| **Type Safety** | ⭐⭐⭐⭐ | Good use of Rust's type system |
| **Ergonomics** | ⭐⭐⭐ | Generally good, some friction points |
| **Consistency** | ⭐⭐⭐ | Room for improvement in naming/error handling |
| **Abstraction Quality** | ⭐⭐⭐ | Solid foundation, some trait complexity |
| **Documentation** | ⭐⭐⭐⭐ | Good doc comments, could use more examples |

**Verdict:** The API design is **professionally competent** with clear Rust idioms. Main improvements needed are in consistency (error handling, naming conventions) and reducing cognitive load around trait bounds.

---

## Next Steps

1. Create follow-up issues for each high-priority recommendation
2. Audit public API surface for error type consistency
3. Document scalar trait relationships in module-level docs
4. Consider RFC process for larger refactoring proposals
